generic.go

```diff
@@ -95,21 +74,75 @@ func NewGeneric(config *Config, queue workqueue.RateLimitingInterface, workers i
 				Help:      fmt.Sprintf("Histogram measuring the time it took to process a %s", &groupKind),
 			},
 		)
+		// Extra api data
+		requestCount := prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: fmt.Sprintf("processed_%s_count", objectName),
+				Help: fmt.Sprintf("Cumulative number of %s processed", &groupKind),
+			},
+			[]string{"url", "method", "status"},
+		)
+		requestTime := prometheus.NewHistogram(
+			prometheus.HistogramOpts{
+				Name: fmt.Sprintf("processed_%s_time", objectName),
+				Help: fmt.Sprintf("Number of seconds each request to %s takes", &groupKind),
+			},
+		)
+
 		allMetrics := []prometheus.Collector{
 			objectProcessCount, objectProcessTime,
+			requestCount, requestTime,
 		}
 		for _, metric := range allMetrics {
 			if err := constructorConfig.Registry.Register(metric); err != nil {
 				return nil, errors.WithStack(err)
 			}
 		}
 
-		holders[descr.Gvk] = Holder{
-			Cntrlr:             iface,
-			ZapNameField:       descr.ZapNameField,
-			ReadyForWork:       readyForWork,
-			objectProcessCount: objectProcessCount,
-			objectProcessTime:  objectProcessTime,
+		constructed, err := constr.New(
+			constructorConfig,
+			&Context{
+				ReadyForWork: func() {
+					close(readyForWork)
+				},
+				Middleware:  addMiddleware(requestCount, requestTime),
+				Informers:   informers,
+				Controllers: controllers,
+				WorkQueue:   queueGvk,
+			},
+		)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to construct controller for GVK %s", descr.Gvk)
+		}
+
+		inf, ok := informers[descr.Gvk]
+		if ok {
+			inf.AddEventHandler(&GenericHandler{
+				Logger:       gvkLogger,
+				WorkQueue:    queueGvk,
+				ZapNameField: descr.ZapNameField,
+			})
+		}
+
+		if constructed.Interface != nil {
+			controllers[descr.Gvk] = constructed.Interface
+
+			holders[descr.Gvk] = Holder{
+				Cntrlr:             constructed.Interface,
+				ZapNameField:       descr.ZapNameField,
+				ReadyForWork:       readyForWork,
+				objectProcessCount: objectProcessCount,
+				objectProcessTime:  objectProcessTime,
+			}
+		}
+
+		if constructed.Server != nil {
+			servers[descr.Gvk] = ServerHolder{
+				Server:       constructed.Server,
+				ZapNameField: descr.ZapNameField,
+				requestCount: requestCount,
+				requestTime:  requestTime,
```

Comment:
It looks like you're missing the "ReadyForWork" field for ServerHolder. Was that supposed to be there?

Reply:
I don't think we need it in servers, do we? For controllers it allows the controller to do slow initialization before any work is handled to it. Hm, maybe servers also need to be able to do something like this. We can leave it out for now and add it when/if we have a real use case? Or just add it.

Reply:
I think @nickhiggs left it in the ServerHolder struct